src/loki/search.cc

```diff
@@ -658,8 +658,23 @@ struct bin_handler_t {
         correlated.edges.erase(new_end, correlated.edges.end());
       }
 
+
+      // Store filtered edges for retry in case find path with correlated.edges failed.
+
+      // We will also add penalty to the score of filtered edges, so as to handle the
+      // case low score filtered_edges VS high score good heading edges during retry.
+      // We use max score of the correlated.edgesfor penalty.
+      float penalty = 0;
+      for (const auto &edge : correlated.edges) {
+        penalty = std::max(edge.score, penalty);
+      }
+      for (auto &filtered_edge : filtered) {
+        filtered_edge.score = filtered_edge.score + penalty;
+        correlated.filtered_edges.push_back(std::move(filtered_edge));
+      }
+
```

Comment:
@kevinkreiser Please see the updated PR as requested, I keep the ```add penalty logic``` here to avoid having to write same logic twice in ```thor_worker_t::get_path``` for both origin and destination.